chain/chain/src/resharding/trie_state_resharder.rs

```diff
@@ -0,0 +1,273 @@
+    }
+}
+
+/// TrieStateResharder is responsible for handling state resharding operations.
+pub struct TrieStateResharder {
+    runtime: Arc<dyn RuntimeAdapter>,
+    /// Controls cancellation of background processing.
+    pub handle: ReshardingHandle,
+    /// Configuration for resharding.
+    resharding_config: MutableConfigValue<ReshardingConfig>,
+}
+
+impl TrieStateResharder {
+    pub fn new(
+        runtime: Arc<dyn RuntimeAdapter>,
+        handle: ReshardingHandle,
+        resharding_config: MutableConfigValue<ReshardingConfig>,
+    ) -> Self {
+        Self { runtime, handle, resharding_config }
+    }
+
+    // Processes one batch of a trie state resharding and updates the status,
+    // also persisting the status to the store.
+    fn process_batch_and_update_status(
+        &self,
+        status: &mut TrieStateReshardingStatus,
+    ) -> Result<(), Error> {
+        let batch_size = self.resharding_config.get().batch_size.as_u64() as usize;
+        let batch_delay = self.resharding_config.get().batch_delay.unsigned_abs();
+        let Some(child) = status.children.first_mut() else {
+            // No more children to process.
+            return Ok(());
+        };
+
+        // Sleep between batches in order to throttle resharding and leave some resource for the
+        // regular node operation.
+        std::thread::sleep(batch_delay);
+        let _span = tracing::debug_span!(
+            target: "resharding",
+            "TrieStateResharder::process_batch_and_update_status",
+            parent_shard_uid = ?status.parent_shard_uid,
+            child_shard_uid = ?child.shard_uid,
+        );
+
+        let mut store_update = self.runtime.store().store_update();
+        let next_key = next_batch(
+            self.runtime.get_tries(),
+            child.shard_uid,
+            child.state_root,
+            child.next_key.clone(),
+            batch_size,
+            &mut store_update.trie_store_update(),
+        )?;
+
+        if let Some(metrics) = &child.metrics {
+            metrics.inc_processed_batches();
+        }
+        if let Some(next_key) = next_key {
+            child.next_key = next_key;
+        } else {
+            // No more keys to process for this child shard.
+            // TODO(resharding): Remove the shard UID mapping from the store.
+            // store_update.trie_store_update().delete_shard_uid_mapping(child.shard_uid);
+            status.children.remove(0);
+        };
+
+        // Commit the changes to the store, along with the status.
+        if status.done() {
+            store_update.delete(DBCol::Misc, TRIE_STATE_RESHARDING_STATUS_KEY);
+        } else {
+            store_update.set(
+                DBCol::Misc,
+                TRIE_STATE_RESHARDING_STATUS_KEY,
+                &borsh::to_vec(status)?,
+            );
+        }
+        store_update.commit()?;
+        Ok(())
+    }
+
+    fn load_status(&self) -> Result<Option<TrieStateReshardingStatus>, Error> {
```

Comment:
Why not load status as part of TrieStateResharder::new() function and store it as opt member of TrieStateResharder? I suspect we only need to write the values after that in each batch, maybe do some quick asserts in `start_resharding_blocking`

Reply:
If we do it in new then new would have to deal with StorageError (which doesn't have a way to deal with it in ReshardingActor::new so I think it's fine as is.